functorch/experimental/cond.py

```diff
@@ -151,7 +152,21 @@ def cond_python_dispatcher(*args):
     return cond(*args)
 
 
+@cond.py_impl(DispatchKey.FuncTorchDynamicLayerFrontMode)
+def cond_functionalize(pred, true_fn, false_fn, inputs):
+    functional_true_fn = functionalize(true_fn)
+    functional_false_fn = functionalize(false_fn)
+
+    _ = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.FuncTorchDynamicLayerFrontMode))
+
+    if pred:
+        return functional_true_fn(*inputs)
```

Comment:
The error you ran into, "RuntimeError: !isFunctionalTensor(tensor) INTERNAL ASSERT FAILED", is due to how functionalization works. We can resolve that error.

At a high level, the way functionalize(f)(x) works is that it wraps x in a FunctionalTensorWrapper and then calls f on the wrapped tensor. The output of f(FunctionalTensorWrapper) is another FunctionalTensorWrapper, and functionalize will unwrap the output to return a regular Tensor.

Now, the error you're running into is because `inputs` are FunctionalTensorWrapper, and functionalization will yell if it tries to construct a FunctionalTensorWrapper(FunctionalTensorWrapper(Tensor)).

## The solution

Pseudocode:
```py
unwrapped_inputs = unwrap_functional_tensors(inputs)
unwrapped_outputs = functional_true_fn(unwrapped_inputs)
return wrap_in_functional_tensor(unwrapped_outputs)
```
You can find the right helper functions in torch/csrc/functorch/init.cpp and/or functorch/_src/eager_transforms.py

Also, instead of using `functionalize(true_fn)`, it might be better to start out with an fx-traced version of it. Something like:

```py
def cond_functionalize(pred, true_fn, false_fn, inputs):
    _ = ExcludeDispatchKeyGuard(...)
    functional_true_fn = make_fx(functionalize(true_fn))(*inputs)
    functional_false_fn = make_fx(functionalize(false_fn))(*inputs)
    return cond(pred, functional_true_fn, functional_false_fn, inputs)
```
(I'm not completely sure that `functionalize(true_fn)` will "just work" due to how functionalization cannot be nested, and using make_fx to inline the functionalization could solve that problem)